Assets/BossRoom/Scripts/Shared/Net/UnityServices/Lobbies/LobbyAsyncRequests.cs

```diff
@@ -0,0 +1,339 @@
+﻿using System;
+using System.Collections.Generic;
+using BossRoom.Scripts.Shared.Infrastructure;
+using BossRoom.Scripts.Shared.Net.UnityServices.Infrastructure;
+using Unity.Multiplayer.Samples.BossRoom;
+using Unity.Services.Authentication;
+using Unity.Services.Lobbies.Models;
+
+namespace BossRoom.Scripts.Shared.Net.UnityServices.Lobbies
+{
+    /// <summary>
+    /// An abstraction layer between the direct calls into the Lobby API and the outcomes you actually want. E.g. you can request to get a readable list of
+    /// current lobbies and not need to make the query call directly.
+    /// </summary>
+    public class LobbyAsyncRequests : IDisposable
+    {
+        private float m_heartbeatTime = 0;
+        private readonly LobbyAPIInterface m_LobbyApiInterface;
+        private readonly UpdateRunner m_SlowUpdate;
+        private readonly LocalLobby m_LocalLobby;
+        private readonly LobbyUser m_LocalLobbyUser;
+        private const float k_heartbeatPeriod = 8; // The heartbeat must be rate-limited to 5 calls per 30 seconds. We'll aim for longer in case periods don't align.
+
+        [Inject]
+        public LobbyAsyncRequests(UpdateRunner slowUpdate, LobbyAPIInterface lobbyAPIInterface, LocalLobby localLobby, LobbyUser localLobbyUser)
+        {
+            m_LobbyApiInterface = lobbyAPIInterface;
+            m_SlowUpdate = slowUpdate;
+            m_LocalLobby = localLobby;
+            m_LocalLobbyUser = localLobbyUser;
+
+            m_RateLimitQuery = new RateLimitCooldown(1.5f, slowUpdate);
+            m_RateLimitJoin = new RateLimitCooldown(3f, slowUpdate);
+            m_RateLimitQuickJoin = new RateLimitCooldown(10f,  slowUpdate);
+            m_RateLimitHost = new RateLimitCooldown(3f, slowUpdate);

```

Comment:
did we give the feedback that a lot of the content of this class should be handled by the lobby's SDK? This seems like a lot of boilerplate users shouldn't have to touch